### PR TITLE
[FEATURE] Weapon/Armor Leveling System (#23)

### LINKED ISSUE #23: [FEATURE] Weapon/Armor Levelling System
**Weapon and Armor Leveling** will provide XP based on mob kills, with different XP amounts for different types of mobs.

- **XP Distribution:** Passive mobs grant 10 XP, and bosses grant `50 XP`. This base XP can be adjusted based on the mob's level. For instance, a level 48 passive mob could have its XP multiplied accordingly.

- **Weapon Leveling:** Each weapon level increases damage by `0.1`.

- **Armor Leveling:** Each armor level increases armor by `0.1` and boosts damage reduction based on a predefined maximum. For example, if the config sets a maximum damage reduction of `75%` and each level grants 1% reduction, you would reach the max reduction well before hitting `500 levels`. Beyond that, damage reduction would cap at `75%`, but armor would continue to increase, reaching +50 armor at level 500.

**TODO**

- [x] Make XP goals from mobs as described above and add XP scaling for each level on both armor and weapons.
- [x] Specify what each level gives to both armor and weapons EX. Each level gives 0.1 attack damage to weapons and 0.1 armor and a percentage of damage reduction to armor.
- [x] Add configs to allow users to specify how much each mob type gives in terms of XP, how much armor and damage reduction each level gives, how much attack damage each level gives, what the max level should be, and the level scaling like the amount of XP between each level, and to turn off the feature.
- [x] Make it so weapons in the wild have random weapon levels, and so does armor. Make it a range of levels per specified dimension. EX. Overworld Range 1-50, Nether Range 70-90, etc

### PR DESCRIPTION
**Pull request by RedstoneWizard08**
Closes #23 
WIP DO NOT MERGE

### CODE REVIEW

File: build.gradle
Comment (naomieow) [inline build.gradle:1]: Please don't include formatting changes in an non-formatting PR.

File: src/main/java/com/bibireden/playerex/mixin/ItemStackMixin.java [hunk 1]
```diff
@@ -99,9 +112,10 @@ public <T extends LivingEntity> void preventBreak(int amount, T entity, Consumer
     public void removeBrokenOnRepair(int damage, CallbackInfo ci) {
         if (!PlayerEX.CONFIG.getItemBreakingEnabled()) return;
 
-        ItemStack stack = (ItemStack)(Object)this;
+        ItemStack stack = (ItemStack) (Object) this;
         if (PlayerEXUtil.isBroken(stack) && damage < stack.getDamageValue()) {
             CompoundTag tag = stack.getTag();
+            assert tag != null;
```
Comment (naomieow) [inline src/main/java/com/bibireden/playerex/mixin/ItemStackMixin.java]: We shouldn't throw errors during runtime and this also doesn't belong in this PR

File: src/main/java/com/bibireden/playerex/mixin/ItemStackMixin.java [hunk 2]
```diff
@@ -142,7 +156,8 @@ public void preventArmour(EquipmentSlot slot, CallbackInfoReturnable<Multimap<At
 
     @Unique
     private String playerex$value(double e, Map.Entry<Attribute, AttributeModifier> entry, AttributeModifier modifier) {
-        if (modifier.getOperation() != AttributeModifier.Operation.ADDITION) return ItemStack.ATTRIBUTE_MODIFIER_FORMAT.format(e);
+        if (modifier.getOperation() != AttributeModifier.Operation.ADDITION)
```
Comment (naomieow) [inline src/main/java/com/bibireden/playerex/mixin/ItemStackMixin.java]: Formatting, and we don't use this form of `if` statement - either use a one line as it was previously or surround with curly braces to improve readabiliy.

File: src/main/java/com/bibireden/playerex/mixin/ItemStackMixin.java
Comment (naomieow) [inline src/main/java/com/bibireden/playerex/mixin/ItemStackMixin.java:1]: Formatting again.

**Review by naomieow — CHANGES_REQUESTED** [review]
Please do not include formatting changes in a non-formatting PR.

**Review by naomieow — DISMISSED** [review]

**Comment by naomieow** [comment]
Please set up GPG/SSH signing for your commits

**Comment by RedstoneWizard08** [comment]
> Please do not include formatting changes in a non-formatting PR.

Ok, will do.

**Comment by RedstoneWizard08** [comment]
> Please set up GPG/SSH signing for your commits
I'll work with Poke on that, as I don't know how. I've only ever used git over HTTPS.

**Comment by naomieow** [comment]
> I'll work with Poke on that, as I don't know how. I've only ever used git over HTTPS.

You can sign your commits with GPG when using HTTPS auth - git just also allows you to sign with SSH as well. I personally use SSH auth and GPG signing. [Here](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) is the GitHub documentation for setting up GPG signing.

**Comment by RedstoneWizard08** [comment]
btw @naomieow I can prob work on this again tomorrow, I've been busy with homework until now

**Comment by naomieow** [comment]
> btw @naomieow I can prob work on this again tomorrow, I've been busy with homework until now

awesome, am currently working on getting basic armour levelling working right now

**Review by cassie-jfs — APPROVED** [review]
lgtm

**Pull request merged**